CMakeLists.txt

```diff
@@ -0,0 +1,61 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+
+project(
+  CPM.cmake
+  VERSION 0.37.0
+  LANGUAGES NONE
+)
+
+include(CMakePackageConfigHelpers)
+configure_package_config_file(
+  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPM.cmakeConfig.cmake.in"
+  "${CMAKE_CURRENT_BINARY_DIR}/CPM.cmakeConfig.cmake"
+  INSTALL_DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/CPM.cmake"
+)
+write_basic_package_version_file(
+  "${CMAKE_CURRENT_BINARY_DIR}/CPM.cmakeConfigVersion.cmake" COMPATIBILITY SameMinorVersion
+                                                                     ARCH_INDEPENDENT
+)
+
+# For consumer using FetchContent
+if(NOT ${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
+  # Check if the user download the repo with git and not in an official release
+  if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git" AND IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/.git")
```

Comment:
I've recently found out about this approach:
[`.gitattributes`](https://github.com/scikit-build/scikit-build-core/blob/c4469f6c600cd45aa5ef62a534f453184c63098d/.gitattributes) and [`.git_archival.txt`](https://github.com/scikit-build/scikit-build-core/blob/main/.git_archival.txt) (this name is arbitrary).

Basically whenever Github creates the tar ball, it will change the content of `.git_archival.txt` with appropriate values. E.g.:
- [Tagged version](https://github.com/scikit-build/scikit-build-core/archive/refs/tags/v0.2.0.tar.gz): `describe-name: v0.2.0`
- [Branch version](https://github.com/scikit-build/scikit-build-core/archive/refs/heads/main.tar.gz): `describe-name: v0.2.0-9-gc4469f6`

Let's go with this approach. We don't need to worry about the versions before this, because you can't use `ExternalProject` on those versions.

Reply:
To test locally, use:
```console
$ git archive --format=tar HEAD --output=test.tar.gz
```

Reply:
@TheLartians As an alternative, would you agree to such approach, i.e.:
- Read `.git_archival.txt` before `project()`
- Reformat the version to be cmake compliant
- Use this variable as the version number

I have never tried this in a project, but could be very elegant if it can be implemented. Things to confirm:
- [ ] cmake accepts this format
- [ ] `CPM_MAJOR_VERSION` and so on are extractable
- [ ] The commit is included in the version and can be retrieved

Reply:
Relevant PR [here](https://github.com/flagarde/CPM.cmake/pull/1). For now this approach works because we don't actually build anything. Basically all of this is now compressed to
```cmake
# Check if there is anything after the allowed format of `vX.Y.Z`, e.g. `vX.Y.Z-rc1`, `vX.Y.Z-9-gc449f6`, etc.
if (GIT_DESCRIBE MATCH "[v]?[0-9\\.]*.+")
  set(CPM_RELEASE FALSE)
else ()
  set(CPM_RELEASE TRUE)
endif ()
```

Reply:
This may, or may not be useful: https://github.com/BareCpper/Version.cmake/blob/38de0dfb1686cfcc0578414587f06e6bcd7415c1/CMake/Version.cmake#L42